toolchain/check/impl_lookup.cpp

```diff
@@ -451,65 +451,38 @@ auto LookupImplWitness(Context& context, SemIR::LocId loc_id,
-
-  // NOTE: Do not retain this reference to the SpecificInterface obtained from a
-  // SpecificInterfaceId. Doing impl lookup does deduce which can do more impl
-  // lookups, and impl lookup can add a new SpecificInterface to the store which
-  // can reallocate and invalidate any references held in here.
-  auto query_specific_interface =
-      context.specific_interfaces().Get(query_specific_interface_id);
-
-  bool query_is_concrete = [&] {
-    if (!query_self_const_id.is_concrete()) {
+static auto QueryIsConcrete(Context& context, SemIR::ConstantId self_const_id,
+                            SemIR::SpecificInterface& specific_interface)
+    -> bool {
+  if (!self_const_id.is_concrete()) {
+    return false;
+  }
+  if (!specific_interface.specific_id.has_value()) {
+    return true;
+  }
+  auto args_id =
+      context.specifics().Get(specific_interface.specific_id).args_id;
+  for (auto inst_id : context.inst_blocks().Get(args_id)) {
+    if (!context.constant_values().Get(inst_id).is_concrete()) {
       return false;
     }
-    if (!query_specific_interface.specific_id.has_value()) {
-      return true;
-    }
-    auto args_id =
-        context.specifics().Get(query_specific_interface.specific_id).args_id;
-    for (auto inst_id : context.inst_blocks().Get(args_id)) {
-      if (!context.constant_values().Get(inst_id).is_concrete()) {
-        return false;
-      }
-    }
-    return true;
-  }();
-
-  // When the query is a concrete FacetValue, we want to look through it at the
-  // underlying type to find all interfaces it implements. This supports
-  // conversion from a FacetValue to any other possible FacetValue, since
-  // conversion depends on impl lookup to verify it is a valid type change. See
-  // https://github.com/carbon-language/carbon-lang/issues/5137. We can't do
-  // this step earlier than inside impl lookup since:
-  // - We want the converted facet value to be preserved in
-  //   `FindWitnessInFacet()` to avoid looking for impl declarations.
-  // - The constant self value may be modified during constant evaluation as a
-  //   more specific value is found.
-  if (auto facet_value = context.insts().TryGetAs<SemIR::FacetValue>(
-          context.constant_values().GetInstId(query_self_const_id))) {
-    query_self_const_id =
-        context.constant_values().Get(facet_value->type_inst_id);
   }
+  return true;
+}
 
-  struct CandidateImpl {
-    SemIR::ImplId impl_id;
-    SemIR::InstId loc_inst_id;
-    TypeStructure type_structure;
-  };
+struct CandidateImpl {
+  SemIR::ImplId impl_id;
+  SemIR::InstId loc_inst_id;
 
-  auto query_type_structure = BuildTypeStructure(
-      context, context.constant_values().GetInstId(query_self_const_id),
-      query_specific_interface);
+  // Used for sorting the candidates to find the most-specialized match.
+  TypeStructure type_structure;
+};
 
-  llvm::SmallVector<CandidateImpl> candidate_impl_ids;
+static auto CollectCandidateImplsForQuery(
```

Comment:
For a function comment, maybe:

```suggestion
// Returns the list of candidates impls for lookup to select from.
static auto CollectCandidateImplsForQuery(
```

Reply:
Done